feat: add support for a unified hpc-stack and jedi-stack

slawrence87544 opened the pull request:
body: This modification of hpc-stack enables support for compiles
of both noaa-emc/gsi and jcsda-internal/soca-science as well as
the ufs weather model.  This branch also adds a stack config which
includes compile of cmake as well as it seems like compile of
the jedi repos goes more smoothly when cmake is compiled vs when
it is untarred from a pre-built package.

slawrence87544 committed:
feat: add support for a unified hpc-stack and jedi-stack

body: This modification of hpc-stack enables support for compiles
of both noaa-emc/gsi and jcsda-internal/soca-science as well as
the ufs weather model.  This branch also adds a stack config which
includes compile of cmake as well as it seems like compile of
the jedi repos goes more smoothly when cmake is compiled vs when
it is untarred from a pre-built package.

edwardhartnett commented:
Can you please start an issue for this work in which you describe what you are doing?

slawrence87544 committed:
refactor: revert udunits to develop branch and modify stack_jedi

body: Revert changes to udunits and use what is in the develop
branch.  Modify stack_jedi.yaml accordingly and set the udunits
version back to 2.2.26.

slawrence87544 commented:
I'm putting one in now, thanks for the reminder.
-Steve

On Thu, Jan 21, 2021 at 12:02 PM Edward Hartnett <notifications@github.com>
wrote:

> Can you please start an issue for this work in which you describe what you
> are doing?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/hpc-stack/pull/145#issuecomment-764867849>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANWW65GONJ3XGJKOIPYCK63S3B24VANCNFSM4WNLIPLA>
> .
>

slawrence87544 commented:
Please refer to issue #146 which is the reason for this pull request.

On Thu, Jan 21, 2021 at 2:32 PM Steve Lawrence <lawrence.steve.m@gmail.com>
wrote:

> I'm putting one in now, thanks for the reminder.
> -Steve
>
> On Thu, Jan 21, 2021 at 12:02 PM Edward Hartnett <notifications@github.com>
> wrote:
>
>> Can you please start an issue for this work in which you describe what
>> you are doing?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/NOAA-EMC/hpc-stack/pull/145#issuecomment-764867849>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ANWW65GONJ3XGJKOIPYCK63S3B24VANCNFSM4WNLIPLA>
>> .
>>
>

kgerheiser commented:
Looks good to me.

slawrence87544 committed:
Merge branch 'develop' into feature/add-support-for-jedi-soca-science

kgerheiser commented:
@aerorahul any input on this?

I tried it out and built it just fine.

kgerheiser commented on libs/build_netcdf.sh:
```diff
@@ -63,7 +63,7 @@ LDFLAGS1="-L$HDF5_ROOT/lib"
 [[ $enable_pnetcdf =~ [yYtT] ]] && LDFLAGS4="-L$PNETCDF_ROOT/lib"
 if [[ ${STACK_netcdf_shared:-} != [yYtT] ]]; then
-  LDFLAGS1+=" -lhdf5_hl -lhdf5"
+  LDFLAGS1+=" -lhdf5 -lhdf5_hl -lz"
```
Was there a reason to swap these?

I think it should be `-lhdf5_hl -lhdf5`. The more specific library comes first.

slawrence87544 replied:
This is now reverted to the develop branch of noaa-emc/hpc-stack.

aerorahul commented on libs/build_netcdf.sh:
```diff
@@ -63,7 +63,7 @@ LDFLAGS1="-L$HDF5_ROOT/lib"
 [[ $enable_pnetcdf =~ [yYtT] ]] && LDFLAGS4="-L$PNETCDF_ROOT/lib"
 if [[ ${STACK_netcdf_shared:-} != [yYtT] ]]; then
-  LDFLAGS1+=" -lhdf5_hl -lhdf5"
+  LDFLAGS1+=" -lhdf5 -lhdf5_hl -lz"
```
why `lz`? Isn't that captured in L67 with the `Extra libraries`?
Also, what if hdf5 was built without zlib?

slawrence87544 replied:
This is now reverted to the develop branch of noaa-emc/hpc-stack

aerorahul commented on libs/build_netcdf.sh:
```diff
@@ -63,7 +63,7 @@ LDFLAGS1="-L$HDF5_ROOT/lib"
 [[ $enable_pnetcdf =~ [yYtT] ]] && LDFLAGS4="-L$PNETCDF_ROOT/lib"
 if [[ ${STACK_netcdf_shared:-} != [yYtT] ]]; then
-  LDFLAGS1+=" -lhdf5_hl -lhdf5"
+  LDFLAGS1+=" -lhdf5 -lhdf5_hl -lz"
```
I have a feeling that some compilers read linkers left to right and some other way.
Is that true?

slawrence87544 replied:
That could be, yes, however I was never able to independently verify that.

slawrence87544 replied:
At any rate, I have reverted both the hdf5 and netcdf library builds to what is found in the develop branch of noaa-emc/hpc-stack.

aerorahul commented on libs/build_netcdf.sh:
```diff
@@ -150,10 +152,10 @@ set -x
 if [[ ${STACK_netcdf_shared} =~ [yYtT] ]]; then
   export LIBS=$($prefix/bin/nc-config --libs)
-  export LDFLAGS+=" -L$prefix/lib"
+  export LDFLAGS+=" -L$prefix/lib -lnetcdf -lhdf5_hl -lhdf5 -lz"
```
These should not be required because `nc-config --libs` should be giving everything except the path to the newly build netcdf-c library

slawrence87544 replied:
See above comments.

aerorahul commented on libs/build_netcdf.sh:
```diff
@@ -150,10 +152,10 @@ set -x
 else
   export LIBS=$($prefix/bin/nc-config --libs --static)
-  export LDFLAGS+=" -L$prefix/lib -lnetcdf"
+  export LDFLAGS+=" -L$prefix/lib -lnetcdf -lhdf5 -lhdf5_hl -lz"
```
Should not be necessary.

slawrence87544 replied:
see above comments.

aerorahul commented on libs/build_netcdf.sh:
```diff
@@ -119,6 +119,8 @@ mkdir -p build && cd build
              --enable-netcdf-4 \
              --disable-doxygen \
+             --enable-parallel4 \
+             --disable-large-file-tests \
```
why disable large file tests? I would think we need them if we turn `MAKE_CHECK=Y` in the `config.sh`

slawrence87544 replied:
see above comments.

aerorahul commented on libs/build_netcdf.sh:
```diff
@@ -110,7 +110,7 @@ software=$name-"c"-$version
 mkdir -p build && cd build
 
-[[ ${STACK_netcdf_shared} =~ [yYtT] ]] || shared_flags="--disable-shared"
+[[ ${STACK_netcdf_shared} =~ [yYtT] ]] && shared_flags="--enable-shared --enable-static" || shared_flags="--disable-shared --enable-static"
```
so are we building shared and static at all times?
If so, why not add that to the `configure` instead of duplicating `--enable-static` here for both if and else.

slawrence87544 replied:
see above comments, but all netcdf and hdf5 build script changes are reverted to noaa-emc/hpc-stack develop branch.

aerorahul commented on libs/build_hdf5.sh:
```diff
@@ -60,9 +60,9 @@ software=$name-$(echo $version | sed 's/\./_/g')
+[[ -z $mpi ]] || extra_conf="--enable-parallel --enable-unsupported --enable-build-mode=production"
 
-[[ $enable_shared =~ [yYtT] ]] || shared_flags="--disable-shared --enable-static --enable-static-exec"
+[[ $enable_shared =~ [yYtT] ]] && shared_flags="--enable-shared --enable-static" || shared_flags="--disable-shared --enable-static --enable-static-exec"
```
if we want to build static at all times, `--enable-static` should move to the `configure` line.

slawrence87544 replied:
see above.

aerorahul commented on libs/build_hdf5.sh:
```diff
@@ -60,9 +60,9 @@ software=$name-$(echo $version | sed 's/\./_/g')
 mkdir -p build && cd build
 
-[[ -z $mpi ]] || extra_conf="--enable-parallel --enable-unsupported"
+[[ -z $mpi ]] || extra_conf="--enable-parallel --enable-unsupported --enable-build-mode=production"
```
I don't know what `--enable-build-mode=production` is, but that is fine to add.

slawrence87544 replied:
see above.

aerorahul commented on config/stack_jedi.yaml:
```diff
@@ -0,0 +1,263 @@
+cmake:
```
how is `stack_jedi` different that `stack_noaa`?
I understand a few (all) are `build: YES`, but we need that on NOAA too (all except wcoss).
So, rather than adding a `stack_jedi`, can we work with `stack_noaa`?
Thoughts?

slawrence87544 replied:
The main difference here from stack_noaa.yaml is building hdf5 and netcdf as shared libraries.  However, after re-checking the CMakeLists.txt file for the noaa-emc/gsi build, I'll try to add the directives FIND_HDF5_HL and FIND_HDF5 to see if that helps.  From memory, those directives were getting pre-empted by the cmake/Modules/platforms/Generic.cmake script in the master branch of the noaa-emc/gsi repo.

aerorahul reviewed (changes requested):
See comments.

slawrence87544 committed:
fix: reverting hdf5 and netcdf build scripts to develop branch

body: also adding cdo build to stack_jedi.yaml and setting builds
of hdf5 and netcdf to shared in stack_jedi.yaml until the generic
platform cmake for the gsi is updated to prevent preemption of
the hdf5 and hdf5_hl find routines.

slawrence87544 committed:
refactor: remove extra lines at end of udunits, hdf5, and netcdf

body: It looks like I accidentally included an extra line at the
eof of build_udunits.hs, build_hdf5.sh, and build_netcdf.sh.  This
commit is removing them.

aerorahul reviewed (approved):
looks good.

slawrence87544 committed:
Merge remote-tracking branch 'upstream/develop' into feature/add-support-for-jedi-soca-science

slawrence87544 committed:
fix: adding ncio build to stack_jedi.yaml
